lib/wakes/external_redirect_mapper.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+module Wakes
+  MAX_REDIRECTS = 10
+
+  class ExternalRedirectMapper
+    def initialize(url)
+      @redirecting_uris = []
+      @uri = URI(url)
+    end
+
+    def resource
+      resolve_target&.resource
+    end
+
+    private
+
+    # rubocop:disable MethodLength
+    def resolve_target
+      target_location = Wakes::Location.find_by(:host => host(@uri), :path => path(@uri))
+
+      redirects = 0
+      while target_location.nil? && expanded_url(@uri).present? && redirects < MAX_REDIRECTS
+        @redirecting_uris << @uri
+        target_location = Wakes::Location
+                          .find_by(:host => host(expanded_url(@uri)), :path => path(expanded_url(@uri)))
+                          &.resource&.canonical_location
+        @uri = URI(expanded_url(@uri))
+        redirects += 1
+      end
+      setup_redirects(target_location)
+      target_location
+    end
+    # rubocop:enable MethodLength
+
+    def setup_redirects(target_location)
+      return if target_location.nil?
+      @redirecting_uris
```

Comment:
@benhutton Because Twitter shortens everything to `t.co`, but then behind those we also have vanity `dsr.gd` URLs. If you accept that it's useful to have a graph of short URLs to resources, then you probably want both.